File: UVa/Python/1048.py
#Nicole Molineros Sánchez
#Código: 8946616
#Como miembro de la comunidad académica de la Pontificia Universidad Javeriana Cali me comprometo a seguir los más altos estándares de integridad académica.

from sys import stdin
from collections import deque
from heapq import heappop, heappush

# Usamos diccionarios para no tener problema con el n de la ciudad
adj = {} 
PrecioTiquete = {}
Iniciales = {}
######################################################################

visitados = {} # key: Estado y value = ints
padre = {} # key: Estado y value = Estado

def BFS(ruta):
    # Entrada: Recibimos como entrada una cola (ruta) 
    # Salida: Se realiza el State Space Search hasta encontrar un 
    # Estado en el cual la ruta esté totalmente vacia
    # por lo que se puede decir que se visitó todo el itinerario, 
    # entonces se retorna el estado (tupla) donde encontramos esa ruta vacia.
    global adj, PrecioTiquete, Iniciales, padre, visitados

    visitados = {}
    padre = {}
    heap = []
    inicial = ruta.popleft()

    rutaTuple = tuple(ruta)
    Estado = (0, inicial, -1, rutaTuple  )
    heappush(heap, Estado)
    visitados[ Estado ] = 1
    padre[Estado] = Estado

    while len(heap) != 0 :

        Estadou = heappop(heap)
        
        preciou = Estadou[0]
        u = Estadou[1]
        tiqueteu = Estadou[2]
        rutaTuple = Estadou[3]

        rutau = deque(rutaTuple)

        if len(rutau) == 0: 
            #Acá es donde se verifica que se haya llegado a un estado en el cual la ruta es vacia
            # O sea que ese estado sea la respuesta y la retornemos
            return Estadou

        for v, tiqv in adj[u]:
            preciov = PrecioTiquete.get(tiqv)
            rutav = deque()
            rutav = deque(rutau)

            if tiqueteu == tiqv:
                # Caso 1
                # Si u y v estan en el mismo tiquete, me muevo
                if len(rutau) > 0:
                    if rutau[0] == v: #  Si se cumple la ruta 
                        rutav.popleft()

                rutaTuple = tuple(rutav)
                Estadov = (preciou, v, tiqueteu, rutaTuple)


                if visitados.get(Estadov) != 1:
                    visitados[Estadov] = 1
                    padre[Estadov] = Estadou
                    heappush(heap, Estadov)

            elif Iniciales.get((u, tiqv)) == 1:
                # Caso 2
                # Si u y v no estan en el mismo tiquete, y v es una ciudad inicial en su tiquete
                if len(rutau) > 0:
                    if rutau[0] == v: #  Si se cumple la ruta 

                        rutav.popleft()
                
                rutaTuple = tuple(rutav)
                Estadov = (preciou + preciov, v, tiqv, rutaTuple)

                if visitados.get(Estadov) != 1:
                    visitados[Estadov] = 1
                    padre[Estadov] = Estadou
                    heappush(heap, Estadov)
    return None


def main():
    l = 0
    lines = stdin.readlines() # Se lee toda la entrada
    cases = 0
    global adj, PrecioTiquete, Iniciales, padre, visitados
    while l < len(lines) - 1: # Se itera la entrada (cada caso), como finaliza con un 0 entonces no se lee la ultima linea
        cases += 1
        adj = {}
        PrecioTiquete = {}
        Iniciales = {}

        nT = int(lines[l].strip()) # Guardamos el numero de tiquetes en nT
        l += 1
        for nTiquete in range(nT):
            tiquete = lines[l].strip().split() # Guardamos la linea y la separamos en espacios en la lista "tiquete" 
            l += 1
            precio = int(tiquete[0])
            
            u = int(tiquete[2]) 

            Iniciales[(u, nTiquete)] = 1 # Marcamos a la primera ciudad u como ciudad inicial en el tiquete nTiquete
            PrecioTiquete[nTiquete] = precio # Guardamos el precio del tiquete nTiquete 
            
            i = 3 # Iniciamos en 3, porque ya leimos el precio del tiquete, el numero de ciudades y la primera ciudad
            while i < len(tiquete):
                v = int(tiquete[i]) 
                if u not in adj: # inicializamos adj[u]  
                    adj[u] = []
                if v not in adj: # inicializamos adj[v]  
                    adj[v] = []
                adj[u].append( (v, nTiquete) ) # Ingresamos el hijo del nodo u, como una tupla (hijo, id tiquete)
                u = v
                i += 1

        nQ = int(lines[l].strip())
        l += 1
        
        for nquery in range(nQ):
            query = lines[l].strip().split()
            l += 1
            ruta = deque()

            i = 1
            while i < len(query):
                ruta.append(int(query[i]))
                i += 1
                
            ans = ()
            ans = BFS(ruta)

            if(ans != None):
                route = []
                node = ans

                if(node in padre):
                    while( padre [node] != node) : 
                        route.append(node)
                        if(node not in padre):
                            break
                        node = padre[node]
                route.append(node)
            
                print("Case {}, Trip {}: Cost = {}".format(cases, nquery + 1, ans[0]))
                print("  Tickets used:", end = "")

                anterior = -1
                j = len(route) - 1
                while j >= 0:
                    if anterior != route[j][2]:
                        print(" {}".format(route[j][2] + 1), end="" )
                        anterior = route[j][2]
                    j -= 1
                print()
            else:
                print("NO")

main()

File: UVa/C/1048.cpp
#include <bits/stdc++.h>

using namespace std;

struct Estado {
    int NodoActual, TiqueteActual, Precio, ruta;

    bool operator<(const Estado &other) const{
        return tie(Precio, NodoActual, TiqueteActual, ruta) < tie(other.Precio, other.NodoActual ,other.TiqueteActual ,other.ruta);
    }
    bool operator>(const Estado &other) const{
        return tie(Precio, NodoActual, TiqueteActual, ruta) > tie(other.Precio, other.NodoActual ,other.TiqueteActual ,other.ruta);
    }

    bool operator!=(const Estado &other) const{
        return (Precio != other.Precio) || 
                (TiqueteActual != other.TiqueteActual) || 
                (NodoActual != other.NodoActual) ||
                (ruta != other.ruta);
    }
    bool operator==(const Estado &other) const{
        return (Precio == other.Precio) && 
                (TiqueteActual == other.TiqueteActual) && 
                (NodoActual == other.NodoActual) &&
                (ruta == other.ruta);
    }
};



int n;
vector<vector<pair<int, int> > > adj;
vector<int> PrecioTiquete, rutaVec;
vector<vector<int>> Iniciales;
map<Estado, int> visitados;
map<Estado, Estado> padre;


Estado BFS(int s){
    int i, j, k, u, v, pesov, costou, inicial, tiqueteu, tiquetev, rutau, rutav;
    Estado Estadotemp, Estadou, Estadov;
    priority_queue<Estado, vector<Estado>, greater<Estado> > cola;


    visitados = map<Estado, int> ();
    padre = map<Estado, Estado> ();


    inicial = s;

    Estadotemp.NodoActual = inicial;
    Estadotemp.TiqueteActual = -1;
    Estadotemp.Precio = 0;
    Estadotemp.ruta = 1;
    cola.push(Estadotemp);
    visitados[Estadotemp] = 1;
    padre[Estadotemp] = Estadotemp;


    while(!cola.empty() ){

        Estadou = cola.top();
        costou = cola.top().Precio;
        u = cola.top().NodoActual;
        rutau = cola.top().ruta;
        tiqueteu = cola.top().TiqueteActual;
        cola.pop();

        if(rutau == rutaVec.size() - 1 ) return Estadou;


        for(j = 0; j < adj[u].size(); ++j){
            
            v = adj[u][j].first;
            pesov = PrecioTiquete[adj[u][j].second];
            tiquetev = adj[u][j].second;

            rutav = rutau;
            if(rutau < rutaVec.size() -1 )
                if(rutaVec[rutau] == v){ /* Si se cumple la ruta  */
                    rutav++;
                }

            if( tiqueteu == tiquetev ){
            /*Caso 1*/
            /* Si u y v estan en el mismo tiquete, me muevo*/
                //cout  << rutau.front() << " a " << v << endl;
                
                Estadov.NodoActual = v;
                Estadov.TiqueteActual = tiqueteu;
                Estadov.Precio = costou;
                Estadov.ruta = rutav;

                if( visitados[Estadov] == 0){
                    // cout << u << " A " << v << endl; 
                    visitados[Estadov] = 1;
                    padre[Estadov] = Estadou;                    
                    cola.push(Estadov);
                    
                }
            }
            else if( Iniciales[u][tiquetev] == 1){
            /* Caso 2 */
            /* Si u y v no estan en el mismo tiquete, y v es una ciudad inicial en su tiquete*/ 
                // cout  << rutau.front() << " a " << v << endl; 

                Estadov.NodoActual = v;
                Estadov.TiqueteActual = tiquetev;
                Estadov.Precio = costou + pesov;
                Estadov.ruta = rutav;

                if( visitados[Estadov] == 0){
                    //cout << u << " B " << v << endl; 
                    visitados[Estadov] = 1;
                    padre[Estadov] = Estadou;                    
                    cola.push(Estadov);
                }
                

            }

        }

    }
    return Estadou;
}

int main(){
    int m, i, aux1, aux2, peso, NT, cases, anterior, maximo, nciudades;
    vector<int> temp;
    Estado ans;
    vector< Estado > path;
    queue<int> ruta;
    Estado nodo;
    cases = 0;
    cin >> NT;
    map<int, int> ciudades;

    while(NT != 0){
        cases++;
        adj = vector<vector<pair<int, int> > >(50000);
        PrecioTiquete = vector<int> (30);
        Iniciales = vector<vector<int>> (30, vector<int>(30, 0));
        visitados.clear();
        ciudades.clear();
        nciudades = 0;


        for (int i = 0; i < NT; i++){
            cin >> peso >> n;
            temp = vector<int>(0);
            for (int j = 0; j < n; j++){
                cin >> aux1;

                if (ciudades.find(aux1) == ciudades.end()){
                    ciudades[aux1] = nciudades;
                    nciudades += 1;
                }

                if(j == 0 ){
                    Iniciales[ciudades[aux1]][i] = 1;
                }
                temp.push_back(ciudades[aux1]);
            }
            PrecioTiquete[i] = peso;
            if(temp.size() != 0)
                for (int j = 0; j < temp.size() - 1; j++){
                    adj[temp[j]].push_back(make_pair( temp[j + 1], i));
                }
        }
        
        cin >> n;

        for (int q = 1; q <= n; q++){
            rutaVec = vector<int>();
            cin >> m;

            for (int i = 0; i < m; i++){
                cin >> aux1;
                rutaVec.push_back(ciudades[aux1]);
            }
            rutaVec.push_back(0);

    
            ans = BFS(rutaVec[0]);

            path = vector< Estado > ();
            nodo = ans;

            while( nodo != padre[nodo] ){
                path.push_back(nodo);
                nodo = padre[nodo];
            }
            path.push_back(nodo);

            printf("Case %d, Trip %d: Cost = %d\n", cases, q, ans.Precio);
            printf("  Tickets used:");

            anterior = -1;
            for (int i = path.size() - 1; i >= 0; i--){
                if(anterior != path[i].TiqueteActual){
                    printf(" %d", (path[i].TiqueteActual + 1));
                    anterior = path[i].TiqueteActual;
                }
            }
            printf("\n");

        }

        cin >> NT;
    }

    return 0;
}
